### DISCUSSION TITLE
Payment request vs Invoice confusion

### DISCUSSION THREAD

**Discussion by Ou7law007** [btcpayserver/btcpayserver#3971] [General]
It seems like it's possible for an end user to make a payment using either. It also seems like invoices automatically create payment requests, right?

After reading both docs, I came to the following conclusions:

1. Invoices never expire??. Payment requests can have an expiration date. Quoting the docs for Invoices: _must be paid within a defined time interval at a fixed exchange rate._ but there's no option to set an expiration date for invoices!
2. Invoices create a payment request automatically (when paid?) and the opposite is not true.
3. Invoices require a buyer email address whereas the email address in the payment requests form is for updates on the payment (seller's email address should be provided).
4. The API supports both but the webhooks only support invoices. Does this mean, that using the payment request API will not trigger any webhooks?

I feel like I'm missing something that answers all my questions!

There's this from Google on PayPal specifically:

> If you are wondering what's the difference between “Money Request” or “Invoice”, here is the answer for you. Money request is mainly used for personal purposes like getting payment from friends or family. The invoice is for business purposes.

Is this true here as well?

This means that invoices are a bit more formal and complicated (provide more options) and should be used for business purposes (since they also trigger the webhooks), whereas payment requests are just a "throw away" that may be used if one doesn't care about accounting etc... and just wants to receive a quick payment with no records etc...

One last thing, that's not really related: How would one generally test webhooks on a local machine? Is there other way since localhost doesn't work and I cannot do port forwarding any way? I need to try and receive the requests on my machine to see how to work with the secret key & handle the backend logic after every event, but without creating a server.

Thank you

**Comment by pavlenex** [comment]
It seems you're mixing invoices and payment request, when you say invoice it's a payment request and vice-versa.

Best it to just generate both and see what they do.

- This is payment request https://pay.pavlenex.com/payment-requests/d91bebef-69d4-49d1-9bc3-c0ab96387b96
- This is an invoice https://pay.pavlenex.com/i/3YdLf1Y5TrzV8Wd6ASMAdT

1. > Invoices never expire??. Payment requests can have an expiration date

No it's the exact opposite. Invoices expire, payment requests don't.

2. > Invoices create a payment request automatically 

No it's the opposite. Payment requests when generates invoice(s)

3. Invoices require a buyer email address whereas the email address in the payment requests form is for updates on the payment 

Invoices do not require an email address it's optional thing you can set up in your store settings. Email addresses in payment requests need to be fixed and improved int he future, it's not funcitonal yet, see #3579 

4. Can't answer this one, don't know maybe @bolatovumar or @ndeet can help.

**Reply by Ou7law007** [reply]
> No it's the exact opposite. Invoices expire, payment requests don't.

But there's an "Expiration Date" option for payment requests and there isn't one for invoices! I also just checked an invoice I had created and it says "State	Expired"

> No it's the opposite. Payment requests when generates invoice(s)

My invoice also created a payment request automatically, once it was paid!

Now, I'm extra confused...

**Reply by pavlenex** [reply]
There's one for invoices, in store settings. Of course, an invoice can expire. Check docs for "invoice statuses".

**Reply by pavlenex** [reply]
> But there's an "Expiration Date" option for payment requests and there isn't one for invoices! I also just checked an invoice I had created and it says "State Expired"

CC @NicolasDorier I guess this is correct, so people are now even more confused, I think "expiration" isn't a proper way to distinguish between a payment request and an invoice.

**Reply by Ou7law007** [reply]
I'm still super confused!!

How bad is it, if I were to use invoices ONLY while completely ignoring payment requests' existence, is this sufficient to receive payments from users & bookkeeping?

**Reply by pavlenex** [reply]
Yes, invoices are the core of BTCPay, payment request is just a layer built on top of it to satisfy a particular use case, for example client invoicing. The naming is very unfortunate for both, and you're not the only one confused, however it's so deep into the code that renaming these would be a significant effort that may break things.

**Reply by Ou7law007** [reply]
A short comparison between the two + an explanation to why this is the case (in the docs) should do.

Either way thank you for answering.

And thanks to everyone who's working on this project <3

**Reply by Kukks** [reply]
Invoices are for short term interactions with a customer ~15mins
Payment requests are for long term interactions with a customer ~days

A payment request is way to request money from a customer that does not need to be paid immediately. When the customer wants to pay, they go on the URL, click on "pay" and an invoice is generated. If the invoice is paid, the payment request is also paid. If the invoice is not paid and expires, the payment request stays pending, until the user repeats the flow successfully. 

The idea is that an invoice will lock in the rate of the invoice currency to the payment method on creation, while a payment request has no currency rate until an invoice is created when the user wishes to pay.

**Reply by Ou7law007** [reply]
That makes sense. This must be in the docs, in my opinion. I could be completely ignorant, though, since I don't see anyone asked this before.

Thanks

**Reply by dennisreimann** [reply]
@pavlenex and I talked about that we should strive for eliminating that confusion all together by making both use cases available via just payment requests. (Also see #3780)

Even as is is right now, you can already have payment requests that behave like invoices: by configuring it with a fixed amount in the desired crypto currency plus an expiry date. 

Invoice as the underlying entity needs to stay and can be exposed via the API though, so that we keep the flexibility for those who know that distinction. 

Imho we should think more about how to go into that direction, as this is a common confusion new users deal with and from what I get as feedback, most users use payment requests and associate that behaviour with what they think of as "invoice".

**Reply by Kukks** [reply]
> @pavlenex and I talked about that we should strive for eliminating that confusion all together by making both use cases available via just payment requests. (Also see #3780)
> 
> Even as is is right now, you can already have payment requests that behave like invoices: by configuring it with a fixed amount in the desired crypto currency plus an expiry date.
> 
> Invoice as the underlying entity needs to stay and can be exposed via the API though, so that we keep the flexibility for those who know that distinction.
> 
> Imho we should think more about how to go into that direction, as this is a common confusion new users deal with and from what I get as feedback, most users use payment requests and associate that behaviour with what they think of as "invoice".

Payment requests are much heavier performance-wise.

**Comment by ndeet** [comment]
Hey @Ou7law007 

> One last thing, that's not really related: How would one generally test webhooks on a local machine? Is there other way since localhost doesn't work and I cannot do port forwarding any way? I need to try and receive the requests on my machine to see how to work with the secret key & handle the backend logic after every event, but without creating a server.

You have a few options here. 
1) you can check the API docs where you see the payload that gets sent by BTCPay, e.g.:
https://docs.btcpayserver.org/API/Greenfield/v1/#tag/Webhooks/paths/InvoicePaymentSettled/post

2) If you want to see what real payloads sent by BTCPay you can use a site like https://beeceptor.com/ where you get your own subdomain that records all the data it get sent to and on BTCPay server you register a webhook to sent to this site. You will see the payload and the header sent by BTCPay like here:
![image](https://user-images.githubusercontent.com/1136761/179210010-14ae839e-10c0-4b82-bb33-476bdf0d8c24.png)

3) On the invoice details page you can also see the fired webhooks and their payload when you click on the webhook ID:
![image](https://user-images.githubusercontent.com/1136761/179210416-abbffac9-e9ff-4742-b0b6-739f93e90813.png)

With the captured data from 2) you can use the payload and BTCPay-Sig header then in an HTTP-Client like [Postman](https://www.postman.com/) or [Insomnia](https://insomnia.rest/products/insomnia) for local development and sent the payload to your app on your local machine.

Alternatively you can expose your local app to the world and sent the webhooks directly to your localhost by using something like [localtunnel](https://localtunnel.github.io/www/) or [ngrok](https://ngrok.com/).

Also not sure what language you use but you might find some useful examples in our PHP lib on how to validate webhooks payloads: 
https://github.com/btcpayserver/btcpayserver-greenfield-php/tree/master/examples

**Comment by Ou7law007** [comment]
> you can check the API docs where you see the payload that gets sent by BTCPay, e.g.:

I was trying to implement the validation properly, which needs a real request AFAIK because it depends on the body, right?
sha256=HMAC256(UTF8(webhook's secret), body)

Either way, I went for ngrok; it took 2 minutes to setup and worked surprisingly fine.

Thank you very much, I really appreciate you.

**Reply by ndeet** [reply]
Yes, you can calculate that hash yourself with the payload but a bit complicated. Therefore I mentioned beeceptor which also logs the `BTCPay-Sig` header for the payload which is is the hash. But not sure if the payload comes with whitespaces and nicely formatted or not, that's why it could fail. But with having ngrok working it should be all good to get the correct payload and you should be able to work the validation out.

You can see here how we validate the Webhook request in PHP: 
https://github.com/btcpayserver/btcpayserver-greenfield-php/blob/master/src/Client/Webhook.php#L215

Great it works for you. 👍
